lib/ClangImporter/SwiftDeclSynthesizer.cpp

```
@@ -1737,17 +1741,27 @@ SubscriptDecl *SwiftDeclSynthesizer::makeSubscript(FuncDecl *getter,
       getterImpl->getClangNode());
   subscript->copyFormalAccessFrom(getterImpl);
 
-  AccessorDecl *getterDecl =
-      AccessorDecl::create(ctx, getterImpl->getLoc(), getterImpl->getLoc(),
-                           AccessorKind::Get, subscript,
-                           /*async*/ false, SourceLoc(),
-                           /*throws*/ false, SourceLoc(),
-                           /*ThrownType=*/TypeLoc(), bodyParams, elementTy, dc);
+  // Use 'address' or 'mutableAddress' accessors for non-copyable
+  // types that are returned indirectly.
+  bool useAddress =
+      rawElementTy->getAnyPointerElementType() && elementTy->isNoncopyable();
+
+  AccessorDecl *getterDecl = AccessorDecl::create(
+      ctx, getterImpl->getLoc(), getterImpl->getLoc(),
+      useAddress ? AccessorKind::Address : AccessorKind::Get, subscript,
+      /*async*/ false, SourceLoc(),
+      /*throws*/ false, SourceLoc(),
+      /*ThrownType=*/TypeLoc(), bodyParams,
+      useAddress ? elementTy->wrapInPointer(PTK_UnsafePointer) : elementTy, dc);
   getterDecl->copyFormalAccessFrom(subscript);
-  getterDecl->setImplicit();
+  if (!useAddress)
+    getterDecl->setImplicit();
```

Should we mark this getter as implicit unconditionally?

---

If we mark an `Address` or `MutableAddress` accessor as implicit, then in `AbstractStorageDecl::getOpaqueAccessor` we return `nullptr` instead of returning the accessor, which causes the compiler to crash. So I think that these kind of accessors are never supposed to be implicit, though I'm not sure what's the reasoning behind this. We also use this trick in `SwiftDeclSynthesizer::makeDereferencedPointeeProperty`